src/peft/tuners/glora/model.py

```diff
@@ -0,0 +1,232 @@
+            return
+
+        model_config_dict = self.model.config.to_dict() if hasattr(self.model.config, "to_dict") else self.model.config
+        current_config = self._prepare_glora_config(config, model_config_dict)
+        self.peft_config[adapter_name] = current_config
+
+        self._find_and_replace(adapter_name)
+
+        mark_only_glora_as_trainable(
+            self.model, bias=current_config.bias if hasattr(current_config, "bias") else "none"
+        )
+
+        if hasattr(current_config, "inference_mode") and current_config.inference_mode:
+            _freeze_adapter(self.model, adapter_name)
+
+    def _check_target_module_exists(self, glora_config: GLoraConfig, key: str) -> bool:
+        if isinstance(glora_config.target_modules, str):
+            return bool(re.fullmatch(glora_config.target_modules, key))
+        elif isinstance(glora_config.target_modules, list):
+            return any(key.endswith(target_key) for target_key in glora_config.target_modules)
+        return False
+
+    def _create_new_module(self, glora_config: GLoraConfig, adapter_name: str, target: nn.Module) -> GLoraLinear:
+        bias = hasattr(target, "bias") and target.bias is not None
+        if not isinstance(target, nn.Linear):
+            raise ValueError(
+                f"Target module {target} is not a nn.Linear layer, which is required for GLORA replacement."
+            )
+
+        in_features, out_features = target.in_features, target.out_features
+        kwargs_glora = {"r": glora_config.r}
+
+        new_module = GLoraLinear(adapter_name, in_features, out_features, bias=bias, **kwargs_glora)
+        return new_module
+
+    def _find_and_replace(self, adapter_name: str):
+        glora_config = self.peft_config[adapter_name]
+        is_target_modules_in_base_model = False
+        key_list = [key for key, _ in self.model.named_modules()]  # Cache keys
+
+        for key in key_list:
+            if not self._check_target_module_exists(glora_config, key):
+                continue
+
+            is_target_modules_in_base_model = True
+            parent, target, target_name = _get_submodules(self.model, key)
+
+            if isinstance(target, GLoraLinear):
+                warnings.warn(
+                    f"Module {key} is already a GLoraLinear. Skipping replacement for new adapter '{adapter_name}'. Multiple GLORA adapters on the same layer might need explicit support in GLoraLinear."
+                )
+            elif isinstance(target, nn.Linear):
+                new_module = self._create_new_module(glora_config, adapter_name, target)
+                self._replace_module(parent, target_name, new_module, target)
+
+        if not is_target_modules_in_base_model:
+            raise ValueError(
+                f"Target modules {glora_config.target_modules} not found in the base model. "
+                f"Please check the target modules and try again."
+            )
```

Comment:
This functionality is already implemented in `BaseTuner.inject_adapter`. I think `_find_and_replace` as well as `add_adapter` can be removed entirely?